csep/core/regions.py

```diff
@@ -753,3 +755,495 @@ def tight_bbox(self):
+        # print('inside If, Current Quad key ', quadk)
+        # print('Length of Quadkey ', len(quadk))
+        # # print('Num of Eqs ', num_eqs)
+
+        _create_tile(quadk + '0', threshold, zoom, lon, lat, qk, num)
+
+        _create_tile(quadk + '1', threshold, zoom, lon, lat, qk, num)
+
+        _create_tile(quadk + '2', threshold, zoom, lon, lat, qk, num)
+
+        _create_tile(quadk + '3', threshold, zoom, lon, lat, qk, num)
+
+    else:
+        # print('inside ELSE, Current Quad key ', quadk)
+        # print('Num of Eqs ', num_eqs)
+        #           qk = numpy.append(qk, quadk)
+        qk.append(quadk)
+        #            num = numpy.append(num, num_eqs)
+        num.append(num_eqs)
+        # return quadk
+
+
+def _create_tile_fix_len(quadk, zoom, qk):
+    """
+    ***Alert: This Function uses GLOBAL variable (qk).
+
+        Provides single-resolution quadtree grid. It takes in a starting quadkey (or Quadrant of Globe),
+        then keeps on keeps on dividing it into 4 children unless the maximum zoom-level is achieved
+        Parameters
+        ----------
+        quadk : String
+            0, 1, 2, 3 or any desired starting level of Quad key.
+            zoom : TYPE
+            Length of Quad Key OR Depth of grid.
+
+
+            Returns
+            -------
+            None.
+        """
+
+    if len(quadk) < zoom:
+        #        print('inside If, Current Quad key ', quadk)
+        #        print('Len of QK: ', len(quadk))
+
+        _create_tile_fix_len(quadk + '0', zoom, qk)
+
+        _create_tile_fix_len(quadk + '1', zoom, qk)
+
+        _create_tile_fix_len(quadk + '2', zoom, qk)
+
+        _create_tile_fix_len(quadk + '3', zoom, qk)
+
+    else:
+        # print('inside ELSE, Current Quad key ', quadk)
+        # print('Num of Eqs ', num_eqs)
+        #        qk = numpy.append(qk, quadk)
+        qk.append(quadk)
+
+
+class QuadtreeGrid2D:
+    """
+    Respresents a 2D quadtree gridded region. The class provides functionality to generate multi-resolution or single-resolution quadtree grid.
+    It also enables users to load already available quadtree grird. It also provides functions to query onto an index 2D grid ad maintains mapping
+    between space coordinates and defined polygons and the index into the polygon array.
+
+    Note: It is replica of CartesianGrid2D class but with quadtree approach, with implementation of all the relevant functions required to CSEP1 tests
+
+    """
+
+    def __init__(self, polygons, quadkeys, bounds, name='QuadtreeGrid2d', mask=None):
+        self.polygons = polygons
+        self.quadkeys = quadkeys
+        self.bounds = bounds  # -----
+        self.poly_mask = mask
+        self.name = name
+
+    @property
+    def num_nodes(self):
+        """ Number of polygons in region """
+        return len(self.polygons)
+
+    @property
+    def get_cell_area(self):
+        """ Area of each cell in grid"""
```

Comment:
expand this docstring a little bit and state that it modifies the state of the class. important because getters typically do not. might be better off being called "compute_cell_area"